src/TorchSharp/NN/Module.cs

```diff
@@ -151,7 +153,9 @@ public virtual Module to(DeviceType deviceType, int deviceIndex = -1)
 
                             if (param is not null) {  // This test must come before the Tensor test
                                 if (deviceType != param.device_type || deviceIndex != param.device_index) {
-                                    var p = new Modules.Parameter(param.to(deviceType, deviceIndex), param.requires_grad);
+                                    var t = param.to(deviceType, deviceIndex);
+                                    t.retain_grad();
```

Comment:
Can you comment on the need for `retain_grad` call here? Is there a bug this fixes?

Reply:
Yes, this is the fix to #516